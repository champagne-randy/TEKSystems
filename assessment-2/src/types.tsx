export interface Vacancies {
  adult: number;
  child: number;
}

export interface RoomData {
  name: string;
  key: string;
  label: string;
  vacancies: Vacancies;
}

export interface RoomRequests {
  adult: number;
  child: number;
}

export interface Room extends RoomData {
  requests: RoomRequests;
  isActive: boolean;
  isRequired: boolean;
}

export interface RoomRequestFormState {
  readonly rooms: Room[];
}

export interface RoomActivationPayload {
  event: {
    target: {
      name: string;
      checked: boolean;
    };
  };
}

export interface RoomActivationHandler {
  (payload: RoomActivationPayload): void;
}

export interface RequestsUpdatePayload {
  name: string;
  data: {
    adult?: number;
    child?: number;
  };
}

export interface RequestsUpdateHandler {
  (payload: RequestsUpdatePayload): void;
}

export interface RoomRequestsSubmissionPayload {
  name: string;
  requests: RoomRequests;
}

export interface RoomProps {
  name: string;
  label: string;
  isActive?: boolean;
  toggleRoomActivation: RoomActivationHandler;
  isRequired?: boolean;
  updateRoomRequests: RequestsUpdateHandler;
  vacancies: Vacancies;
  requests: RoomRequests;
}
