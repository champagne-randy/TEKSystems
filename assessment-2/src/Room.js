import React from "react";
import PropTypes from "prop-types";
import Dropdown from "react-dropdown";
import { range } from "lodash";
import { compose, withProps } from "recompose";
import ShouldShow from "react-should-show";
import classNames from "classnames";
import { Validate } from "./utils";
import "react-dropdown/style.css";
import "./Room.scss";

const Room = ({
  name,
  label,
  isActive,
  toggleRoomActivation,
  isRequired,
  requests,
  updateRoomRequests,
  options
}) => (
  <section
    className={classNames("room", {
      room__disabled: !isActive,
      room__valid:
        isActive &&
        (requests.adult.touched || requests.child.touched) &&
        requests.isValid,
      room__invalid:
        isActive &&
        (requests.adult.touched || requests.child.touched) &&
        !requests.isValid
    })}
    data-testid="room"
    role="group"
  >
    <header className="room__header">
      <ShouldShow shouldShow={!isRequired}>
        <input
          type="checkbox"
          name={name}
          id={`${name}__activation`}
          checked={isActive}
          onChange={event => toggleRoomActivation({ event })}
          data-testid="room__header__checkbox"
        />
      </ShouldShow>
      <label htmlFor={`${name}__activation`}>
        <h2>{label}</h2>
      </label>
    </header>
    <main
      className={classNames("room__requests", {
        room__requests__active: isActive
      })}
      data-testid="room__requests"
    >
      <div role="group">
        <h3>
          Adults
          <br />
          (18+)
        </h3>
        <Dropdown
          className="room__requests__dropdown"
          options={options.adult}
          onChange={({ value }) =>
            updateRoomRequests({
              name,
              data: { adult: { value, touched: true } }
            })
          }
          value={{ label: requests.adult.value, value: requests.adult.value }}
          disabled={!isActive}
          data-testid="room__requests_dropdown--adult"
        />
      </div>
      <div role="group">
        <h3>
          Children
          <br />
          (0 - 17)
        </h3>
        <Dropdown
          className="room__requests__dropdown"
          options={options.child}
          onChange={({ value }) =>
            updateRoomRequests({
              name,
              data: { child: { value, touched: true } }
            })
          }
          value={{ label: requests.child.value, value: requests.child.value }}
          disabled={!isActive}
          data-testid="room__requests_dropdown--child"
        />
      </div>
    </main>
    <footer>
      <ShouldShow
        shouldShow={
          (requests.adult.touched || requests.child.touched) &&
          !requests.isValid
        }
      >
        <div role="alert">Please select at least 1 adult or child room</div>
      </ShouldShow>
    </footer>
  </section>
);

Room.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  isActive: PropTypes.bool.isRequired,
  isRequired: PropTypes.bool.isRequired,
  toggleRoomActivation: PropTypes.func.isRequired,
  requests: PropTypes.shape({
    adult: PropTypes.shape({
      value: PropTypes.number.isRequired,
      touched: PropTypes.bool.isRequired
    }).isRequired,
    child: PropTypes.shape({
      value: PropTypes.number.isRequired,
      touched: PropTypes.bool.isRequired
    }).isRequired
  }).isRequired,
  updateRoomRequests: PropTypes.func.isRequired,
  availability: PropTypes.shape({
    adult: Validate.prop.greaterThanZero,
    child: Validate.prop.greaterThanZero
  }).isRequired,
  options: PropTypes.shape({
    adult: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
    child: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired
  }).isRequired
};

export default compose(
  withProps(({ availability }) => ({
    options: {
      adult: range(0, availability.adult + 1),
      child: range(0, availability.child + 1)
    }
  }))
)(Room);
