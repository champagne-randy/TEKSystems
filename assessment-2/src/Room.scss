.room {
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  padding: 10px;
  border-radius: 5px;
  border: solid 5px #e7e7e7;
  background-color: #e7e7e7;
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  justify-content: flex-start;
  align-items: space-around;
  font-size: 9px;
  &__disabled {
    border-color: #cdd0df;
    background-color: #dbdbe3;
  }
  &__valid {
    border: solid 5px green;
  }
  &__invalid {
    border: solid 5px red;
  }
  &__header {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    justify-content: flex-start;
    align-items: center;
    input {
      margin-right: 5px;
    }
  }
  &__requests {
    width: 100%;
    height: fit-content;
    padding-bottom: 10px;
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    justify-content: space-around;
    align-items: flex-start;
    text-align: left;
    &__active {
      background-color: #ffffff;
    }
    div[role="group"] {
      width: 45%;
    }
    &__dropdown {
      .Dropdown-disabled {
        background-color: #f0f0f0;
      }
    }
  }
}
