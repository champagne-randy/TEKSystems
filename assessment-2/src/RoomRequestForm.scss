.room-request-form {
  width: 100%;
  border: none;
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  justify-content: center;
  align-items: flex-start;
  &__room__container {
    width: 100%;
    border: none;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: flex-start;
    align-items: flex-start;
    .room-request-form__room__wrapper {
      display: inline-block;
      width: 24%;
      flex-basis: 75px;
      margin: 0;
      margin-right: 10px;
      margin-bottom: 20px;
      padding: 0;
      &:last-child {
        margin-right: 0;
        align-self: auto;
      }
    }
  }
  &__submit {
    background-color: #dbdbe3;
  }
}
