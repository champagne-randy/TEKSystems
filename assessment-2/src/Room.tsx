import React, { StatelessComponent } from "react";
import Dropdown from "react-dropdown";
import classNames from "classnames";
import { range } from "lodash";
import { RoomProps } from "./interfaces";
import { getDDOptsFromVacancies } from "./utils";
import "react-dropdown/style.css";
import "./Room.scss";

const Room: StatelessComponent<RoomProps> = props => {
  const {
    name,
    label,
    isActive,
    toggleRoomActivation,
    isRequired,
    vacancies,
    requests,
    updateRoomRequests
  } = props;

  const showValidationError =
    isActive &&
    ((requests.adult.touched || requests.child.touched) && !requests.isValid);

  const options = getDDOptsFromVacancies({ vacancies });

  return (
    <section
      className={classNames("room", {
        room__disabled: !isActive,
        room__valid:
          isActive &&
          (requests.adult.touched || requests.child.touched) &&
          requests.isValid,
        room__invalid:
          isActive &&
          (requests.adult.touched || requests.child.touched) &&
          !requests.isValid
      })}
      data-testid="room"
      role="group"
    >
      <header className="room__header">
        {!isRequired && (
          <input
            type="checkbox"
            name={name}
            id={`${name}__activation`}
            checked={isActive}
            onChange={event => toggleRoomActivation({ event })}
            data-testid="room__header__checkbox"
          />
        )}
        <label htmlFor={`${name}__activation`}>
          <h2>{label}</h2>
        </label>
      </header>
      <main
        className={classNames("room__requests", {
          room__requests__active: isActive
        })}
        data-testid="room__requests"
      >
        <div role="group">
          <label htmlFor="room__requests_dropdown--adult">
            <h3>
              Adults
              <br />
              (18+)
            </h3>
          </label>
          <select
            id="room__requests_dropdown--adult"
            className="room__requests__dropdown"
            disabled={!isActive}
            data-testid="room__requests_dropdown--adult"
          >
            {range(1, vacancies.adult + 1).map(value => (
              <option
                key={`room__requests__dropdown__option--adult-${value}`}
                value={value}
              >
                {value}
              </option>
            ))}
          </select>
        </div>
        <div role="group">
          <label htmlFor="room__requests_dropdown--child">
            <h3>
              Children
              <br />
              (0 - 17)
            </h3>
          </label>
          <select
            id="room__requests_dropdown--child"
            className="room__requests_dropdown"
            disabled={!isActive}
            data-testid="room__requests_dropdown--child"
          >
            {range(0, vacancies.child + 1).map(value => (
              <option
                key={`room__requests__dropdown__option--child-${value}`}
                value={value}
              >
                {value}
              </option>
            ))}
          </select>
        </div>
      </main>
      <footer className="room__requests__validation">
        {showValidationError && (
          <div className="room__requests__validation__message" role="alert">
            Please select at least 1 adult or child room
          </div>
        )}
      </footer>
    </section>
  );
};

/*
Room.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  isActive: PropTypes.bool.isRequired,
  isRequired: PropTypes.bool.isRequired,
  toggleRoomActivation: PropTypes.func.isRequired,
  requests: PropTypes.shape({
    adult: PropTypes.shape({
      value: PropTypes.number.isRequired,
      touched: PropTypes.bool.isRequired
    }).isRequired,
    child: PropTypes.shape({
      value: PropTypes.number.isRequired,
      touched: PropTypes.bool.isRequired
    }).isRequired
  }).isRequired,
  updateRoomRequests: PropTypes.func.isRequired,
  vacancies: PropTypes.shape({
    adult: Validate.prop.greaterThanZero,
    child: Validate.prop.greaterThanZero
  }).isRequired,
  options: PropTypes.shape({
    adult: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
    child: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired
  }).isRequired
};
*/

export default Room;
