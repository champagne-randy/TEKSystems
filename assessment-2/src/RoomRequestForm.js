import React from "react";
import PropTypes from "prop-types";
import Room from "./Room";
import { Validate } from "./utils";
import "./RoomRequestForm.scss";

const RoomRequestForm = ({
  rooms,
  activeRooms,
  requestedRooms,
  toggleRoomActivation,
  updateRoomRequests,
  handleFormSubmit
}) => (
  <form className="room-request-form" onSubmit={handleFormSubmit}>
    <div className="room-request-form__room__container" role="group">
      {rooms.map(room => (
        <div key={room.key} className="room-request-form__room__wrapper">
          <Room
            name={room.name}
            label={room.label}
            isActive={room.isActive}
            isRequired={room.isRequired}
            requests={room.requests}
            toggleRoomActivation={toggleRoomActivation}
            updateRoomRequests={updateRoomRequests}
            availability={room.availability}
          />
        </div>
      ))}
    </div>
    <button className="room-request-form__submit" type="submit">
      Submit
    </button>
  </form>
);

RoomRequestForm.propTypes = {
  rooms: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      key: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      availability: PropTypes.shape({
        adult: PropTypes.number.isRequired,
        child: PropTypes.number.isRequired
      }).isRequired,
      requests: PropTypes.shape({
        adult: PropTypes.shape({
          value: PropTypes.number.isRequired,
          touched: PropTypes.bool.isRequired
        }).isRequired,
        child: PropTypes.shape({
          value: PropTypes.number.isRequired,
          touched: PropTypes.bool.isRequired
        }).isRequired,
        valid: PropTypes.bool.isRequired
      }).isRequired,
      isActive: PropTypes.bool.isRequired,
      isRequired: PropTypes.bool.isRequired
    }).isRequired
  ).isRequired,
  toggleRoomActivation: PropTypes.func.isRequired,
  updateRoomRequests: PropTypes.func.isRequired,
  handleFormSubmit: PropTypes.func.isRequired
};

export default RoomRequestForm;
