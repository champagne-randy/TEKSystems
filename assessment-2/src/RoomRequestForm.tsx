import React, { PureComponent, FormEvent } from "react";
import { findIndex } from "lodash";
import styled from "./styled-components";
import Room from "./Room";
import {
  RoomRequestFormState,
  RoomData,
  RoomActivationHandler,
  RequestsUpdateHandler,
  RoomRequestsSubmissionPayload
} from "./types";

const Form = styled("form")`
  width: 100%;
  border: none;
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  justify-content: center;
  align-items: flex-start;
`;

const Fieldset = styled("div")`
  width: 100%;
  border: none;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
  align-items: flex-start;
`;

const Wrapper = styled("div")`
  display: inline-block;
  width: 24%;
  margin: 0;
  margin-right: 10px;
  margin-bottom: 20px;
  padding: 0;
  &:last-child {
    margin-right: 0;
    align-self: auto;
  }
`;

class RoomRequestForm extends PureComponent {
  // TODO:
  // - hook into apollo cache instead of React state?
  readonly state: RoomRequestFormState = { rooms: [] };

  constructor(props: any) {
    super(props);
    this.toggleRoomActivation = this.toggleRoomActivation.bind(this);
    this.updateRoomRequests = this.updateRoomRequests.bind(this);
    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  // TODO:
  // - impl mock rest call with Axios
  // - use async await to demo promise handling
  // - cache form data in Apollo cache
  componentDidMount() {
    const payload = require("./room.data.json");
    this.setState({
      rooms: payload.map((room: RoomData, idx: number) => ({
        ...room,
        requests: {
          adult: 1,
          child: 0
        },
        isActive: idx === 0,
        isRequired: idx === 0
      }))
    });
  }

  toggleRoomActivation: RoomActivationHandler = payload => {
    const { name, checked: isActive } = payload.event.target;
    const { rooms } = this.state;
    const index = findIndex(rooms, { name });
    this.setState({
      rooms: rooms.map((room, idx) => ({
        ...room,
        isActive: isActive ? idx <= index : idx < index
      }))
    });
  };

  updateRoomRequests: RequestsUpdateHandler = payload => {
    const { name, data } = payload;
    this.setState({
      rooms: this.state.rooms.map(room => ({
        ...room,
        requests:
          room.name === name
            ? {
                ...room.requests,
                ...data
              }
            : room.requests
      }))
    });
  };

  // TODO:
  // - cache form data in Apollo cache
  // - impl mock rest call with Axios
  handleFormSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const roomRequests: RoomRequestsSubmissionPayload[] = this.state.rooms
      .filter(room => room.isActive)
      .map(room => ({
        name: room.name,
        requests: room.requests
      }));

    console.log("RoomRequestForm submitted:");
    console.dir(roomRequests);
    alert(
      `RoomRequestForm submitted:\n${JSON.stringify(roomRequests, null, 4)}`
    );
  };

  render() {
    // Note:
    // - I chose to only render component if there's room data
    // - This behavior can easily be changed i.e. to render an empy/disabled state
    if (this.state.rooms.length === 0) {
      return null;
    }

    return (
      <Form onSubmit={this.handleFormSubmit}>
        <Fieldset role="group">
          {this.state.rooms.map(room => (
            <Wrapper key={room.key}>
              <Room
                name={room.name}
                label={room.label}
                isActive={room.isActive}
                isRequired={room.isRequired}
                toggleRoomActivation={this.toggleRoomActivation}
                updateRoomRequests={this.updateRoomRequests}
                vacancies={room.vacancies}
                requests={room.requests}
              />
            </Wrapper>
          ))}
        </Fieldset>
        <button type="submit">Submit</button>
      </Form>
    );
  }
}

export default RoomRequestForm;
