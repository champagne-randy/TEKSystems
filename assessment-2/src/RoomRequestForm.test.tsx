import React from "react";
import { shallow, mount } from "enzyme";
import {
  render,
  fireEvent,
  cleanup,
  waitForElement
} from "react-testing-library";
import RoomRequestForm from "./RoomRequestForm";
import Room from "./Room";
import { RoomRequestFormState } from "./types";
import "jest-dom/extend-expect";

afterEach(cleanup);

describe("<RoomRequestForm />: structure", () => {
  it("renders without crashing", async () => {
    const { container } = await waitForElement(() =>
      render(<RoomRequestForm />)
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("<RoomRequestForm />: behavior", () => {
  it("room activation works as expected", async () => {
    const {
      container,
      getByTestId,
      queryAllByTestId,
      getByText,
      debug
    } = render(<RoomRequestForm />);

    expect(container.firstChild).toMatchSnapshot();

    // expect(getByTestId(/room-1__checkbox/i, { exact: false })).toBeNull;

    const checkbox2: HTMLElement = await waitForElement(
      () => getByTestId(/room-2__checkbox/i, { exact: false }),
      { container }
    );
    const checkbox3: HTMLElement = await waitForElement(
      () => getByTestId(/room-3__checkbox/i, { exact: false }),
      { container }
    );
    const checkbox4: HTMLElement = await waitForElement(
      () => getByTestId(/room-4__checkbox/i, { exact: false }),
      { container }
    );

    // initial form state
    // expect(checkbox1).toBeNull;
    expect(checkbox2).toHaveProperty("checked", false);
    expect(checkbox3).toHaveProperty("checked", false);
    expect(checkbox4).toHaveProperty("checked", false);

    // state after activating 3rd room
    fireEvent.click(checkbox3, { button: 0 });
    // expect(checkbox1).toBeNull;
    expect(checkbox2).toHaveProperty("checked", true);
    expect(checkbox3).toHaveProperty("checked", true);
    expect(checkbox4).toHaveProperty("checked", false);

    // state after deactivating 3rd room
    fireEvent.click(checkbox3, { button: 0 });
    // expect(checkbox1).toBeNull;
    expect(checkbox2).toHaveProperty("checked", true);
    expect(checkbox3).toHaveProperty("checked", false);
    expect(checkbox4).toHaveProperty("checked", false);
  });
});
