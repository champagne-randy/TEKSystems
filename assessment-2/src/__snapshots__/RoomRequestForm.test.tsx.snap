// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomRequestForm />: behavior room activation works as expected 1`] = `null`;

exports[`<RoomRequestForm />: structure renders without crashing 1`] = `null`;
